src/generic_grader/utils/turtle_canvas.py

```diff
@@ -0,0 +1,152 @@
+import inspect
+import turtle
+from io import BytesIO
+from os.path import isfile
+from pathlib import Path
+
+from attrs import evolve
+from PIL import Image, ImageChops
+
+from generic_grader.utils.user import RefUser, SubUser
+
+
+def save_ref_canvas(setup_func, draw_func, args, pen_width, filename, invert):
+    # Setup the turtle and canvas.
+    setup_func()
+
+    # Skip animation
+    turtle.tracer(0)
+
+    turtle.width(pen_width)  # Override reference turtle width.
+
+    # Draw the spiral.
+    draw_func(*args)
+
+    # Save the canvas and reset.
+    turtle.hideturtle()
+    turtle.update()
+    save_canvas(turtle.getcanvas(), filename, invert)
+    turtle.reset()
+
+
+def save_sub_canvas(self, options):
+    o = options
+
+    if o.entries:
+        sol_image_fn = f"sol_{o.entries}.png"
+        sol_image_fn_inv = f"sol_{o.entries}_inv.png"
+    else:
+        sol_image_fn = "sol.png"
+        sol_image_fn_inv = "sol_inv.png"
+
+    # Skip if solution file already exists.
+    if isfile(sol_image_fn) and isfile(sol_image_fn_inv):
+        return
+
+    # Create the reference and student user.
+    ref_options = evolve(o, obj_name="start", entries=())
+    self.ref_start_user = RefUser(self, ref_options)
+    self.student_main_user = SubUser(self, o)
+
+    # Initialize the turtle window using the reference modules start().
+    self.ref_start_user.call_obj()
+
+    # Do we want to force the log limit to be 0? End users may want to have some output and an image created
+    # log_limit = calc_log_limit([])
```

Comment:
Using the reference log here could lead to unexpected behavior, I believe it would be better to just use the value from options, and adjust the default value in options.